returnn/training.py

```diff
@@ -573,3 +573,179 @@ def hash(cls, kwargs):
+        :param key: a key from the learning rate file that is used to sort the models,
+            e.g. "dev_score_output/output_prob"
+        :param index: index of the sorted list to access, 0 for the lowest, -1 for the highest score
+        """
+        self.model_dir = model_dir
+        self.learning_rates = learning_rates
+        self.index = index
+        self.out_epoch = self.output_var("epoch")
+        self.key = key
+
+    def run(self):
+        # this has to be defined in order for "eval" to work
+        def EpochData(learningRate, error):
+            return {"learning_rate": learningRate, "error": error}
+
+        with open(self.learning_rates.get_path(), "rt") as f:
+            text = f.read()
+
+        data = eval(
+            text, {"nan": float("nan"), "inf": float("inf"), "EpochData": EpochData}
+        )
+
+        epochs = list(sorted(data.keys()))
+
+        # some epochs might not have all keys, we require that the last entry has the key, other epochs which
+        # do not have it might be ignored
+        available_keys = data[epochs[-1]]["error"]
+        if self.key not in available_keys:
+            raise KeyError(
+                f"{self.key} is not available in the provided learning_rates file f{self.learning_rates.get_path()}"
+            )
+
+        scores = [
+            (epoch, data[epoch]["error"][self.key])
+            for epoch in epochs
+            if self.key in data[epoch]["error"]
+        ]
+        sorted_scores = list(sorted(scores, key=lambda x: x[1]))
+
+        self.out_epoch.set(sorted_scores[self.index][0])
+
+    def tasks(self):
+        yield Task("run", mini_task=True)
+
+
+class GetBestTFCheckpointJob(GetBestEpochJob):
+    """
+    Returns the best checkpoint given a training model dir and a learning-rates file
+    The best checkpoint will be HARD-linked if possible, so that no space is wasted but also the model not
+    deleted in case that the training folder is removed.
+    """
+
+    def __init__(
+        self, model_dir: tk.Path, learning_rates: tk.Path, key: str, index: int = 0
+    ):
+        """
+
+        :param Path model_dir: model_dir output from a RETURNNTrainingJob
+        :param Path learning_rates: learning_rates output from a RETURNNTrainingJob
+        :param int index: index of the sorted list to access, 0 for the lowest, -1 for the highest score
+        :param str key: a key from the learning rate file that is used to sort the models
+            e.g. "dev_score_output/output_prob"
+        """
+        super().__init__(model_dir, learning_rates, key, index)
+        self._out_model_dir = self.output_path("model", directory=True)
+
+        # Note: checkpoint.index (without epoch number) is only allowed be cause we are symlinking
+        # to a file that contains an epoch number, as RETURNN will resolve symlinks automatically and
+        # then check for the epoch number
```

Comment:
I would reformulate that. There is no reason that RETURNN must behave this way. It's just that way for historical reasons. You could easily also change RETURNN if you would like to.

Reply:
As long as this is not actually merged into RETURNN, I want to keep this as reminder for myself. If you do not want to have this text I can just link the issue.

> There is no reason that RETURNN must behave this way. It's just that way for historical reasons. You could easily also change RETURNN if you would like to.

But I also have to say I do not see how this text is in any way conflicting with those statements.

Reply:
All what you write are technical internal details, only true for current RETURNN, dependent on actual somewhat buggy or strange behavior of RETURNN, and will likely become wrong in the future. From your comment, it reads as if you would always expect RETURNN to behave this way, or a reader of the comment could get this impression.

I would formulate it more in a way like: "Currently RETURNN has the flaw that it parses the epoch number out of the checkpoint filename. This still works if it is a symlink to such a filename. This will likely change in the future but we anyway want to support current and older RETURNN versions."

Reply:
Yes I already wrote a new text